test/integration/good/code-gen/closures/closures1.stan

```diff
@@ -0,0 +1,35 @@
+data {
+    int N;
+    real x;
+    real y[N];
+    vector[N] v;
+}
+transformed data {
+    function
+    real foo(real z, row_vector r) {
+        real rs = sum(r);
+        real ys = sum(y);
+        real vs = sum(v);
+        return z + rs + x + ys + vs;
+    }
+}
+parameters {
+    real p;
+    real pa[N];
+    vector[N] pv;
+}
+model {
+    function
+    real bar(real z, row_vector r) {
+        real rs = sum(r);
+        real ys = sum(y);
+        real vs = sum(v);
+        real pas = sum(pa);
+        real pvs = sum(pv);
+        return z + rs + x + ys + vs + p + pas + pvs;
+    }
```

Comment:
Do you have an example of a failure when there is shadowing in the function and a value is brought in from a higher scope? Like 

```suggestion
    function
    real bar(real z, row_vector r) {
        vector[N] pa = // ... fill this with something
        real rs = sum(r);
        real ys = sum(y);
        real vs = sum(v);
        real pas = sum(pa);
        real pvs = sum(pv);
        return z + rs + x + ys + vs + p + pas + pvs;
    }
```

Reply:
Also can you give the test function names that line up with what they intend to check or validate? It makes it easier when looking across the stan/mir/cpp

Reply:
C++ also allows this in lambdas for `constexpr` defined objects, but that's a very specific type of scoping that's known globally at compile time

Reply:
The Stan language does not allow shadowing. Closures are no exception.

Reply:
Oh good!

Reply:
Would you mind adding an example that takes in no arguments and uses just scoped objects? Reading over the gen'd C++ right now and would be easier to parse through that logic.

Reply:
It won't let me comment on the C++ file because it's too many changes, for review (or maybe just a good idea) it would be good to print those into a separate closure.expected file

Reply:
I'll comment on one here though

```cpp
template <typename T0__, typename T1__, typename T2__, typename T3__,
typename T6__, typename T7__>
stan::promote_args_t<T0__, T1__, stan::value_type_t<T2__>,
stan::value_type_t<T3__>,
T6__, stan::promote_args_t<stan::value_type_t<T7__>>>
bar_L23C9_impl__(const T0__& p, const std::vector<T1__>& pa, const T2__& pv,
                 const T3__& v, const double& x,
                 const std::vector<double>& y, const T6__& z, const T7__& r,
                 std::ostream* pstream__)  ;

template<bool ref__, typename F0__, typename F1__, typename F2__>
class bar_L23C9_cfunctor__ {
  /**
   * When would these be captured by value?
   */
  stan::capture_type_t<F0__, ref__> p;
  stan::capture_type_t<std::vector<F1__>, ref__> pa;
  stan::capture_type_t<Eigen::Matrix<F2__, -1, 1>, ref__> pv;
  const Eigen::Matrix<double, -1, 1>& v;
  stan::capture_type_t<double, ref__> x;
  const std::vector<double>& y;
  public:
  const size_t vars_count__;
  bar_L23C9_cfunctor__(const bar_L23C9_cfunctor__<ref__, F0__, F1__, F2__>&) = default ;
  bar_L23C9_cfunctor__(bar_L23C9_cfunctor__<ref__, F0__, F1__, F2__>&&) = default ;
  bar_L23C9_cfunctor__(const F0__& p__, const std::vector<F1__>& pa__,
                       const Eigen::Matrix<F2__, -1, 1>& pv__,
                       const Eigen::Matrix<double, -1, 1>& v__,
                       const double& x__, const std::vector<double>& y__)
  : p(p__), pa(pa__), pv(pv__), v(v__), x(x__), y(y__),
    vars_count__(count_vars(p__, pa__, pv__, v__, x__, y__)) {}
  template <typename T0__, typename T1__>
  stan::promote_args_t<F0__, F1__, F2__, T0__,
stan::value_type_t<T1__>>
  operator()(std::ostream* pstream__, const T0__& z, const T1__& r)  const 
  {
  return bar_L23C9_impl__(p, pa, pv, v, x, y, z, r, pstream__);
  }
  /**
   * Will any of these be used in places we will also have user defined code?
   *  if not, then no need to double underscore.
   */   
  using captured_scalar_t__ = stan::return_type_t<F0__, F1__, F2__>;
  using ValueOf__ = bar_L23C9_cfunctor__<false, double, double, double>;
  using CopyOf__ = bar_L23C9_cfunctor__<false,
                                        stan::capture_type_t<F0__, false>,
                                        stan::capture_type_t<F1__, false>,
                                        stan::capture_type_t<F2__, false>>;
  /**
   * idt we need these? I think you could just have a function that 
   * returns a tuple of the members like 
   */
   auto get_members() {
     return std::make_tuple(p, pa, pv, v, x, y);
   }
   /**
    * Then on the stan math side we can just run stuff like
    */
    // in math deep_copy_vars.hpp file 
    template <typename F, require_stan_closure_t<F>* = nullptr>
    auto deep_copy_vars(F&& f) {
      return stan::math::for_each([](auto&& x) { return deep_copy_vars(x);}, f.get_members());      
    }
    
  size_t count_vars__() const {
  return vars_count__;
  }
  auto value_of__() const {
  return ValueOf__(eval(value_of(p)), eval(value_of(pa)), eval(value_of(pv)),
                   v, x, y);
  }
  auto deep_copy_vars__() const {
  return CopyOf__(eval(deep_copy_vars(p)), eval(deep_copy_vars(pa)),
                  eval(deep_copy_vars(pv)), v, x, y);
  }
  void zero_adjoints__() {
  stan::math::zero_adjoints(p);
  stan::math::zero_adjoints(pa);
  stan::math::zero_adjoints(pv);
  stan::math::zero_adjoints(v);
  stan::math::zero_adjoints(x);
  stan::math::zero_adjoints(y);
  }
  double* accumulate_adjoints__(double *dest) const {
  return stan::math::accumulate_adjoints(dest, p, pa, pv, v, x, y);
  }
  stan::math::vari** save_varis__(stan::math::vari **dest) const {
  return stan::math::save_varis(dest, p, pa, pv, v, x, y);
  }
  
  
  };
```